File: reference-states/workflows-java/src/test/java/com/example/flow/IOUFlowTests.java
package com.example.flow;

import com.example.contract.SanctionedEntitiesContract;
import com.example.state.SanctionableIOUState;
import com.google.common.collect.ImmutableList;
import net.corda.core.contracts.TransactionVerificationException;
import net.corda.core.flows.FlowSession;
import net.corda.core.flows.NotaryException;
import net.corda.core.identity.Party;
import net.corda.core.node.NetworkParameters;
import net.corda.core.transactions.SignedTransaction;
import net.corda.testing.node.*;
import org.hamcrest.core.IsInstanceOf;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

import java.security.SignatureException;
import java.time.Duration;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

import static java.util.Collections.emptyList;
import static java.util.Collections.emptyMap;
import static net.corda.testing.core.TestConstants.DUMMY_NOTARY_NAME;
import static net.corda.testing.node.internal.InternalTestUtilsKt.findCordapp;

public class IOUFlowTests {

    MockNetwork network;
    StartedMockNode a;
    StartedMockNode b;
    StartedMockNode c;
    StartedMockNode issuer;
    Party issuerParty;

    @Before
    public void setup(){
        MockNetworkParameters param = new MockNetworkParameters(
                false,
                false,
                new InMemoryMessagingNetwork.ServicePeerAllocationStrategy.Random(),
                ImmutableList.of(new MockNetworkNotarySpec(DUMMY_NOTARY_NAME,false)),
                new NetworkParameters(4,emptyList(),10484760, 10484760 * 50, Instant.now(), 1, emptyMap(), Duration.ofDays(30)),
                 ImmutableList.of(findCordapp("com.example.contract"))
        );

        network = new MockNetwork(param);

        a = network.createPartyNode(null);
        b = network.createPartyNode(null);
        c = network.createPartyNode(null);

        issuer = network.createPartyNode(null);
        issuerParty = issuer.getInfo().getLegalIdentities().get(0);
        // For real nodes this happens automatically, but we have to manually register the flow for tests.

        ImmutableList.of(a,b,c ).forEach(node->{
            node.registerInitiatedFlow(IOUIssueFlow.Acceptor.class);
        });
        issuer.registerInitiatedFlow(GetSanctionsListFlow.Acceptor.class);
        network.runNetwork();

    }

    @After
    public void tearDown(){
        network.stopNodes();
    }

    @Rule
    public ExpectedException thrown = ExpectedException.none();
    @Test
    public void dealFailsIfThereIsNoIssuedSanctionsList() throws ExecutionException, InterruptedException, SignatureException {
        thrown.expectCause(IsInstanceOf.<Throwable>instanceOf(TransactionVerificationException.class));
        IOUIssueFlow.Initiator flow = new IOUIssueFlow.Initiator(1, b.getInfo().getLegalIdentities().get(0), issuerParty);
        Future future = a.startFlow(flow);
        network.runNetwork();
        SignedTransaction signedTx = (SignedTransaction) future.get();
        signedTx.verifySignaturesExcept(b.getInfo().getLegalIdentities().get(0).getOwningKey());
        System.out.println(signedTx.getCoreTransaction().outputsOfType(SanctionableIOUState.class).get(0));
    }

    @Test
    public void dealSucceedsWithIssuedSanctions() throws ExecutionException, InterruptedException, SignatureException {
        Future issuanceFlow = issuer.startFlow(new IssueSanctionsListFlow.Initiator());
        network.runNetwork();
        issuanceFlow.get();
        getSanctionsList(a, issuerParty);
        IOUIssueFlow.Initiator flow = new IOUIssueFlow.Initiator(1, b.getInfo().getLegalIdentities().get(0), issuerParty);
        Future future  =  a.startFlow(flow);
        network.runNetwork();
        SignedTransaction signedTx = (SignedTransaction) future.get();
        signedTx.verifySignaturesExcept(b.getInfo().getLegalIdentities().get(0).getOwningKey());

    }


    @Test
    public void dealIsRejectedIfPartyIsSanctioned() throws ExecutionException, InterruptedException {
        thrown.expectCause(IsInstanceOf.<Throwable>instanceOf(TransactionVerificationException.class));
        Future issuanceFlow = issuer.startFlow(new IssueSanctionsListFlow.Initiator());
        network.runNetwork();
        issuanceFlow.get();

        Future updateFuture = issuer.startFlow(new UpdateSanctionsListFlow.Initiator(b.getInfo().getLegalIdentities().get(0)));
        network.runNetwork();
        updateFuture.get();


        IOUIssueFlow.Initiator flow = new IOUIssueFlow.Initiator(1, b.getInfo().getLegalIdentities().get(0), issuerParty);
        Future future = a.startFlow(flow);
        network.runNetwork();

        future.get();
    }

    @Test
    public void dealFailsIfListIsUpdatedAfterCollection() throws ExecutionException, InterruptedException, SignatureException {
        thrown.expectCause(IsInstanceOf.<Throwable>instanceOf(NotaryException.class));
        Future issuanceFlow = issuer.startFlow(new IssueSanctionsListFlow.Initiator());
        network.runNetwork();
        issuanceFlow.get();
        getSanctionsList(a, issuerParty);

        IOUIssueFlow.Initiator iouIssueFlow = new IOUIssueFlow.Initiator(1, b.getInfo().getLegalIdentities().get(0), issuerParty);
        Future iouIssueFuture = a.startFlow(iouIssueFlow);
        network.runNetwork();

        SignedTransaction iouIssueTransaction = (SignedTransaction) iouIssueFuture.get();
        iouIssueTransaction.verifySignaturesExcept(b.getInfo().getLegalIdentities().get(0).getOwningKey());

        Future updateIOUFuture = issuer.startFlow(new UpdateSanctionsListFlow.Initiator(c.getInfo().getLegalIdentities().get(0)));
        network.runNetwork();
        updateIOUFuture.get();

        IOUIssueFlow.Initiator iouReissue = new IOUIssueFlow.Initiator(1, b.getInfo().getLegalIdentities().get(0), issuerParty);
        Future iouReissueFuture = a.startFlow(iouReissue);
        network.runNetwork();

        iouReissueFuture.get();
    }

    @Test
    public void dealSucceedsIfListIsCollectedAgainAfterUpdate() throws ExecutionException, InterruptedException, SignatureException {
        Future issuanceFlow = issuer.startFlow(new IssueSanctionsListFlow.Initiator());
        network.runNetwork();
        issuanceFlow.get();
        getSanctionsList(a, issuerParty);

        IOUIssueFlow.Initiator iouIssueFlow = new IOUIssueFlow.Initiator(1, b.getInfo().getLegalIdentities().get(0), issuerParty);
        Future iouIssueFuture = a.startFlow(iouIssueFlow);
        network.runNetwork();

        SignedTransaction iouIssueTransaction = (SignedTransaction) iouIssueFuture.get();
        iouIssueTransaction.verifySignaturesExcept(b.getInfo().getLegalIdentities().get(0).getOwningKey());

        Future updateIOUFuture = issuer.startFlow(new UpdateSanctionsListFlow.Initiator(c.getInfo().getLegalIdentities().get(0)));
        network.runNetwork();
        updateIOUFuture.get();

        //update on node a only
        Future getUpdatedListAgain = a.startFlow(new GetSanctionsListFlow.Initiator(issuerParty));
        network.runNetwork();
        getUpdatedListAgain.get();

        IOUIssueFlow.Initiator iouDeal2 = new IOUIssueFlow.Initiator(1, b.getInfo().getLegalIdentities().get(0), issuerParty);
        Future iouDeal2Future = a.startFlow(iouDeal2);
        network.runNetwork();
        iouDeal2Future.get();

        IOUIssueFlow.Initiator iouDeal3 = new IOUIssueFlow.Initiator(1, b.getInfo().getLegalIdentities().get(0), issuerParty);
        Future iouDeal3Future = a.startFlow(iouDeal3);
        network.runNetwork();

        iouDeal3Future.get();
    }

    @Test
    public void duringTxResolutionLatestRefIsProvidedToCounterparty() throws ExecutionException, InterruptedException, SignatureException {
        Future issuanceFlow = issuer.startFlow(new IssueSanctionsListFlow.Initiator());
        network.runNetwork();
        issuanceFlow.get();
        getSanctionsList(a, issuerParty);

        IOUIssueFlow.Initiator iouIssueFlow = new IOUIssueFlow.Initiator(1, b.getInfo().getLegalIdentities().get(0), issuerParty);
        Future iouIssueFuture = a.startFlow(iouIssueFlow);
        network.runNetwork();

        SignedTransaction iouIssueTransaction = (SignedTransaction) iouIssueFuture.get();
        iouIssueTransaction.verifySignaturesExcept(b.getInfo().getLegalIdentities().get(0).getOwningKey());

        Future updateIOUFuture = issuer.startFlow(new UpdateSanctionsListFlow.Initiator(c.getInfo().getLegalIdentities().get(0)));
        network.runNetwork();
        updateIOUFuture.get();

        //update on node a only
        Future getUpdatedListAgain = a.startFlow(new GetSanctionsListFlow.Initiator(issuerParty));
        network.runNetwork();
        getUpdatedListAgain.get();

        //take down issuer, so isn't able to provide new list to node b
        issuer.stop();

        IOUIssueFlow.Initiator iouDeal2 = new IOUIssueFlow.Initiator(1, b.getInfo().getLegalIdentities().get(0), issuerParty);
        Future iouDeal2Future = a.startFlow(iouDeal2);
        network.runNetwork();
        iouDeal2Future.get();


    }


    private void getSanctionsList(StartedMockNode node, Party issuerOfSanctions) throws ExecutionException, InterruptedException {
        Future flow = node.startFlow(new GetSanctionsListFlow.Initiator(issuerOfSanctions));
        network.runNetwork();
        flow.get();


    }




}

File: reference-states/workflows-kotlin/src/test/kotlin/com/example/flow/IOUFlowTests.kt
package com.example.flow

import com.example.state.SanctionableIOUState
import com.example.state.SanctionedEntities
import net.corda.core.contracts.TransactionVerificationException
import net.corda.core.flows.NotaryException
import net.corda.core.identity.Party
import net.corda.core.utilities.getOrThrow
import net.corda.testing.common.internal.testNetworkParameters
import net.corda.testing.core.singleIdentity
import net.corda.testing.node.MockNetwork
import net.corda.testing.node.MockNetworkParameters
import net.corda.testing.node.StartedMockNode
import org.junit.After
import org.junit.Before
import org.junit.Test
import java.time.Duration
import java.time.temporal.ChronoUnit

class IOUFlowTests {
    lateinit var network: MockNetwork
    lateinit var a: StartedMockNode
    lateinit var b: StartedMockNode
    lateinit var c: StartedMockNode
    lateinit var issuer: StartedMockNode
    lateinit var issuerParty: Party

    @Before
    fun setup() {
        network = MockNetwork(
            listOf("com.example.contract", "com.example.schema"), MockNetworkParameters(
                networkParameters = testNetworkParameters(
                    minimumPlatformVersion = 4
                )
            )
        )
        a = network.createPartyNode()
        b = network.createPartyNode()
        c = network.createPartyNode()

        issuer = network.createPartyNode()
        issuerParty = issuer.info.legalIdentities.single()
        // For real nodes this happens automatically, but we have to manually register the flow for tests.
        listOf(a, b, c).forEach { it.registerInitiatedFlow(IOUIssueFlow.Acceptor::class.java) }
        listOf(issuer).forEach { it.registerInitiatedFlow(GetSanctionsListFlow.Acceptor::class.java) }
        network.runNetwork()
    }

    @After
    fun tearDown() {
        network.stopNodes()
    }

    @Test(expected = TransactionVerificationException.ContractRejection::class)
    fun `deal fails if there is no issued sanctions list`() {
        val flow = IOUIssueFlow.Initiator(1, b.info.singleIdentity(), issuerParty)
        val future = a.startFlow(flow)
        network.runNetwork()

        val signedTx = future.getOrThrow()
        signedTx.verifySignaturesExcept(b.info.singleIdentity().owningKey)

        println(signedTx.coreTransaction.outputsOfType(SanctionableIOUState::class.java).single())
    }

    @Test
    fun `deal succeeds with issued sanctions`() {
        val issuanceFlow = issuer.startFlow(IssueSanctionsListFlow.Initiator())
        network.runNetwork()
        issuanceFlow.getOrThrow()

        getSanctionsList(a, issuerParty)

        val flow = IOUIssueFlow.Initiator(1, b.info.singleIdentity(), issuerParty)
        val future = a.startFlow(flow)
        network.runNetwork()

        val signedTx = future.getOrThrow()
        signedTx.verifySignaturesExcept(b.info.singleIdentity().owningKey)

        println("@@@@@@@@@@${a.services.vaultService.queryBy(SanctionedEntities::class.java).states.firstOrNull()}")

    }

    @Test(expected = TransactionVerificationException.ContractRejection::class)
    fun `deal is rejected if party is sanctioned`() {
        val issuanceFlow = issuer.startFlow(IssueSanctionsListFlow.Initiator())
        network.runNetwork()
        issuanceFlow.getOrThrow()

        val updateFuture = issuer.startFlow(UpdateSanctionsListFlow.Initiator(b.info.legalIdentities.first()))
        network.runNetwork()
        updateFuture.getOrThrow()

        val flow = IOUIssueFlow.Initiator(1, b.info.singleIdentity(), issuerParty)
        val future = a.startFlow(flow)
        network.runNetwork()

        future.getOrThrow()
    }

    @Test(expected = NotaryException::class)
    fun `deal fails if list is updated after collection`() {
        val issuanceFlow = issuer.startFlow(IssueSanctionsListFlow.Initiator())
        network.runNetwork()
        issuanceFlow.getOrThrow()
        getSanctionsList(a, issuerParty)

        val flow1 = IOUIssueFlow.Initiator(1, b.info.singleIdentity(), issuerParty)
        val future1 = a.startFlow(flow1)
        network.runNetwork()

        val signedTx1 = future1.getOrThrow()
        signedTx1.verifySignaturesExcept(b.info.singleIdentity().owningKey)

        val updateFuture = issuer.startFlow(UpdateSanctionsListFlow.Initiator(c.info.legalIdentities.first()))
        network.runNetwork()
        updateFuture.getOrThrow()

        val flow2 = IOUIssueFlow.Initiator(1, b.info.singleIdentity(), issuerParty)
        val future2 = a.startFlow(flow2)
        network.runNetwork()

        future2.getOrThrow()
    }

    @Test
    fun `deal succeeds if list is collected again after update`() {
        val issuanceFlow = issuer.startFlow(IssueSanctionsListFlow.Initiator())
        network.runNetwork()
        issuanceFlow.getOrThrow()

        getSanctionsList(a, issuerParty)

        val dealFlow1 = IOUIssueFlow.Initiator(1, b.info.singleIdentity(), issuerParty)
        val dealFuture1 = a.startFlow(dealFlow1)
        network.runNetwork()

        val signedTx1 = dealFuture1.getOrThrow()
        signedTx1.verifySignaturesExcept(b.info.singleIdentity().owningKey)

        val updateFuture = issuer.startFlow(UpdateSanctionsListFlow.Initiator(c.info.legalIdentities.first()))
        network.runNetwork()
        updateFuture.getOrThrow()

        //update on node a only
        val getUpdatedListAgain = a.startFlow(GetSanctionsListFlow.Initiator(issuerParty))
        network.runNetwork()
        getUpdatedListAgain.getOrThrow()

        val dealFlow2 = IOUIssueFlow.Initiator(1, b.info.singleIdentity(), issuerParty)
        val dealFuture2 = a.startFlow(dealFlow2)
        network.runNetwork()
        dealFuture2.getOrThrow()

        //perform deal with b again
        val dealFlow3 = IOUIssueFlow.Initiator(1, b.info.singleIdentity(), issuerParty)
        val dealFuture3 = a.startFlow(dealFlow3)
        network.runNetwork()

        dealFuture3.getOrThrow()

    }

    private fun getSanctionsList(node: StartedMockNode, issuerOfSanctions: Party) {
        node.startFlow(GetSanctionsListFlow.Initiator(issuerOfSanctions)).also {
            network.runNetwork()
        }.getOrThrow()
        network.runNetwork()
    }

    @Test
    fun `during tx resoludtion, latest ref state is provided to counterparty`() {
        val issuanceFlow = issuer.startFlow(IssueSanctionsListFlow.Initiator())
        network.runNetwork()
        issuanceFlow.getOrThrow()

        getSanctionsList(a, issuerParty)


        val dealFlow1 = IOUIssueFlow.Initiator(1, b.info.singleIdentity(), issuerParty)
        val dealFuture1 = a.startFlow(dealFlow1)
        network.runNetwork()

        val signedTx1 = dealFuture1.getOrThrow()
        signedTx1.verifySignaturesExcept(b.info.singleIdentity().owningKey)

        val updateFuture = issuer.startFlow(UpdateSanctionsListFlow.Initiator(c.info.legalIdentities.first()))
        network.runNetwork()
        updateFuture.getOrThrow()

        //update on node a only
        getSanctionsList(a, issuerParty)


        val dealFlow2 = IOUIssueFlow.Initiator(1, b.info.singleIdentity(), issuerParty)
        val dealFuture2 = a.startFlow(dealFlow2)
        network.runNetwork()
        dealFuture2.getOrThrow()

        //perform deal with b again
        val dealFlow3 = IOUIssueFlow.Initiator(1, b.info.singleIdentity(), issuerParty)
        val dealFuture3 = a.startFlow(dealFlow3)
        network.runNetwork()

        dealFuture3.getOrThrow()

        //take down issuer, so isn't able to provide new list to node b
        issuer.stop()

        val dealFlow4 = IOUIssueFlow.Initiator(1, a.info.singleIdentity(), issuerParty)
        val dealFuture4 = b.startFlow(dealFlow4)
        network.runNetwork()

        dealFuture4.getOrThrow(timeout = Duration.of(30, ChronoUnit.MINUTES))

    }
}